### PR TITLE
CV2-4386 initial ideas for pushing to async

### PR DESCRIPTION
**Pull request by DGaffney**
## Description

Shifts the main routine (only the main routine, not the other uses) of alegre fingerprinting to use async instead of sync. Since this is happening before the merge for video, only image and audio are effected by this change (we'll incorporate video after this though). We send async requests and receive responses from alegre via alegre_webhook.rb, which then calls back to the alegre module to store the results. We have to store results for confirmed and suggested async queries temporarily in redis, and then run the relating step (note that in `get_similar_items_v2_callback` we only return relations once we are confirmed to have every in-process async request returned in redis).

References: CV2-4386

## How has this been tested?

Extensively tested locally and confirmed to be working. A few missing lines of coverage that I'll address once review is complete

## Things to pay attention to during code review

Lots of new functions in alegre_v2.rb - not a ton of editing in place, so pay attention to the new functions/routines.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [reply app/models/concerns/alegre_v2.rb:258]: Doesn't look like you're using `type` for anything here, Devin.

File: app/models/concerns/alegre_webhooks.rb
Comment (caiosba) [inline app/models/concerns/alegre_webhooks.rb:15]: Please remove this before merging (or use a proper `Rails.logger.info` call)

File: test/models/bot/alegre_3_test.rb
Comment (caiosba) [inline test/models/bot/alegre_3_test.rb:133]: Please remove the :zzz: before merging lol

**Review by caiosba — CHANGES_REQUESTED** [review]
Thanks Devin! I added some early review comments

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [inline app/models/concerns/alegre_v2.rb:365]: @DGaffney, can this method still work as sync for now? I'm thinking about two use cases beyond regular similarity matching that would require further changes to the workflow: tipline search and feed search.

**Review by caiosba — COMMENTED** [review]
Code looks clean, thanks Devin for addressing the comments. I left one question there. We can handle that in a separate ticket once this one is done, if it still works. Let's discuss today in the ML call, but I suggest you also test the tipline use case locally. I'll follow up with you on this.

**Comment by caiosba** [comment]
Also: As usual, let's make sure integration tests pass for this branch before we merge this :)

**Comment by DGaffney** [comment]
@caiosba this is ready for final review!

File: app/models/concerns/smooch_search.rb
Comment (caiosba) [reply app/models/concerns/smooch_search.rb:87]: @DGaffney you're not using `result_id` for anything in that block, is it expected? If so, please rename it to `_result_id`, as per the Code Climate suggestion above.

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [reply app/models/concerns/alegre_v2.rb:4]: @DGaffney , since `media` is defined as an `attr_accessor`, this means that methods `media` (getter), `media=` (setter) are created... but then you define a method called `media` (like a custom getter). If you just want the setter, please use `attr_writer` for `:media` instead of `attr_accessor`.

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [inline app/models/concerns/alegre_v2.rb:37]: Suggestion / non-blocker: You could omit all those `return` words.

File: app/models/bot/alegre.rb
Comment (caiosba) [inline app/models/bot/alegre.rb:402]: Suggestion / non-blocker: You could rewrite as `pm.is_a?(TemporaryProjectMedia)`.

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [inline app/models/concerns/alegre_v2.rb:204]: Suggestion / non-blocker: You could rewrite as `pm.is_a?(TemporaryProjectMedia)`.

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [inline app/models/concerns/alegre_v2.rb:369]: Can you please just add a comment explaining what this line is doing? A lot of Ruby magic here hehe.

File: app/models/concerns/alegre_webhooks.rb
Comment (caiosba) [inline app/models/concerns/alegre_webhooks.rb:25]: I suggest we add the `redis.expire` call under an `ensure` block... this way, even if something goes wrong, we make sure that Redis is going to be cleared (in that case we may need to declare `redis` outside the begin/rescue block).

File: app/models/concerns/smooch_search.rb
Comment (caiosba) [inline app/models/concerns/smooch_search.rb:176]: This doesn't seem to be an error... if we still want to log this situation, let's log as `info` and use our logging pattern (`[Smooch Bot] <Message>`).

File: app/models/concerns/smooch_search.rb
Comment (caiosba) [inline app/models/concerns/smooch_search.rb:178]: This doesn't seem to be an error... if we still want to log this situation, let's log as `info` and use our logging pattern (`[Smooch Bot] <Message>`).

**Review by caiosba — CHANGES_REQUESTED** [review]
Thanks Devin, I left some comments, some of which are only suggestions. Now I'm going to test it locally.

**Comment by caiosba** [comment]
Devin, once you address the comments from my previous review, please re-request a final review. I finished testing it locally and I was able to get text and images to match for items created inside the application and also for tipline searches and feed queries:

Captura de tela de 2024-04-06 19-09-47

Captura de tela de 2024-04-06 19-10-18

:tada:

**Comment by DGaffney** [comment]
@caiosba I've reviewed your notes over the weekend and fixed them all! Take another look when you're ready :)

**Review by caiosba — CHANGES_REQUESTED** [review]
Thanks @DGaffney ! There is one last one here: https://github.com/meedan/check-api/pull/1827#discussion_r1554431665

File: app/models/concerns/smooch_search.rb
Comment (DGaffney) [reply app/models/concerns/smooch_search.rb:87]: on it!

**Comment by DGaffney** [comment]
> Thanks @DGaffney ! There is one last one here: [#1827 (comment)](https://github.com/meedan/check-api/pull/1827#discussion_r1554431665)

Fixed and updated!

**Review by caiosba — APPROVED** [review]
Thanks Devin for addressing all PR comments!

File: app/models/concerns/alegre_v2.rb
Comment (caiosba) [inline app/models/concerns/alegre_v2.rb:424]: @DGaffney can you please notify Sentry if we reach the timeout? This is how you can do it: 

(1) Define a specific class for this error, something like (at the top of the file):

```
class AlegreTimeoutError < StandardError; end
```

(2) Call Sentry:

```
CheckSentry.notify(AlegreTimeoutError.new('Timeout when waiting for async response from Alegre'), params: { project_media: project_media })
```

**Review by caiosba — CHANGES_REQUESTED** [review]

**Review by caiosba — APPROVED** [review]
Devin and I exchanged a few DMs about it - good to go!

**Comment by caiosba** [comment]
*Good to go if tests pass, CC is happy, etc.

**Pull request merged**
